packages/client/src/components/app/forms/RelationshipField.svelte

```diff
@@ -50,55 +36,47 @@
       limit: 100,
     },
   })
-  $: hasFilter = !!filter?.filter(f => !!f.field)?.length
-  $: fetch.update({ filter })
-  $: {
-    options = searchResults ? searchResults : $fetch.rows
-    const nonMatchingOptions = selectedOptions.filter(
-      option => !options.map(opt => opt._id).includes(option._id)
-    )
-    // Append initially selected options if there is no filter
-    // and hasn't already been appended
-    if (!hasFilter) {
-      options = [...options, ...nonMatchingOptions]
-    }
-  }
+
   $: tableDefinition = $fetch.definition
-  $: primaryDisplay = tableDefinition?.primaryDisplay || "_id"
-  $: singleValue = flatten(fieldState?.value)?.[0]
-  $: multiValue = flatten(fieldState?.value) ?? []
+  $: selectedValue = multiselect
+    ? flatten(fieldState?.value) ?? []
+    : flatten(fieldState?.value)?.[0]
   $: component = multiselect ? CoreMultiselect : CoreSelect
   $: expandedDefaultValue = expand(defaultValue)
-  $: debouncedSearch(searchString)
-  $: {
-    if (searching) {
-      isOpen = true
-    }
+  $: primaryDisplay = tableDefinition?.primaryDisplay
+
+  $: initialValues =
+    initialValues ||
+    (primaryDisplay &&
+      fieldState?.value?.map(x => ({
+        _id: x._id,
+        [primaryDisplay]: x.primaryDisplay,
+      })))
+
+  $: allOptions = {
+    ...(allOptions || {}),
+    ...[...($fetch.rows || []), ...(initialValues || [])]?.reduce((p, c) => {
+      p[c._id] = c
+      return p
+    }, {}),
   }
 
-  // Fetch the initially selected values
-  // as they may not be within the first 100 records
-  $: {
-    if (
-      primaryDisplay !== "_id" &&
-      fieldState?.value?.length &&
-      !selectedOptions?.length
-    ) {
-      API.searchTable({
-        paginate: false,
-        tableId: linkedTableId,
-        limit: 100,
-        query: {
-          oneOf: {
-            [`1:${primaryDisplay}`]: fieldState?.value?.map(
-              value => value.primaryDisplay
-            ),
-          },
-        },
-      }).then(response => {
-        const value = multiselect ? multiValue : singleValue
-        selectedOptions = response.rows.filter(row => value.includes(row._id))
-      })
+  $: options = Object.values(allOptions)
+
+  $: fetchRows(searchTerm, primaryDisplay)
+
+  const fetchRows = (searchTerm, primaryDisplay) => {
+    const allRowsFetched =
+      $fetch.loaded &&
+      !Object.keys($fetch.query.string).length &&
+      !$fetch.hasNextPage
+    if (!allRowsFetched) {
+      // Don't request until we have the primary display
+      if (primaryDisplay) {
```

Comment:
Could be combined into a single if statement rather than nested:

`if (!allRowsFetched && primaryDisplay)`